File: facebook/OC/LC/LC/LC721.m
//
//  LC721.m
//  LC
//
//  Created by ULS on 3/7/18.
//  Copyright © 2018 ULS. All rights reserved.
//

#import "LC721.h"
#import "NSMutableArray+Stack.h"

@interface LC721()

- (NSArray *)accountsMerge:(NSArray <NSArray <NSString *> *> *)accounts;

@end

@implementation LC721

/**
 Given a list accounts, each element accounts[i] is a list of strings, where the first element accounts[i][0] is a name, and the rest of the elements are emails representing emails of the account.
 Now, we would like to merge these accounts. Two accounts definitely belong to the same person if there is some email that is common to both accounts. Note that even if two accounts have the same
 name, they may belong to different people as people could have the same name. A person can have any number of accounts initially, but all of their accounts definitely have the same name.
 After merging the accounts, return the accounts in the following format: the first element of each account is the name, and the rest of the elements are emails in sorted order. The accounts
 themselves can be returned in any order.
 */

- (NSArray *)accountsMerge:(NSArray <NSArray <NSString *> *> *)accounts {
    NSMutableDictionary *emailToName = [NSMutableDictionary dictionary];
    NSMutableDictionary<NSString *, NSMutableArray *> *graph = [NSMutableDictionary dictionary];
    for(NSArray <NSString *> *list in accounts) {
        NSString *name = @"";
        for(NSString *element in list) {
            if([name isEqualToString:@""]) {
                name = [element copy];
                continue;
            }
            if([graph objectForKey:element] == NULL) {
                NSMutableArray *emailList = [NSMutableArray array];
                [graph setObject:emailList forKey:element];
            }
            NSMutableArray *emailList = [graph objectForKey:element];
            [emailList addObject:list[1]];
            [graph setObject:emailList forKey:element];
            
            if([graph objectForKey:list[1]] == NULL) {
                NSMutableArray *emailList = [NSMutableArray array];
                [graph setObject:emailList forKey:list[1]];
            }
            NSMutableArray *anotherEmailList = [graph objectForKey:list[1]];
            [anotherEmailList addObject:element];
            [graph setObject:anotherEmailList forKey:list[1]];
            [emailToName setObject:name forKey:element];
        }
    }
    
    [graph enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSMutableArray * _Nonnull obj, BOOL * _Nonnull stop) {
        DLog(@"key:%@", key);
        [obj enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            NSString *email = (NSString *)obj;
            DLog(@"email:%@", email);
        }];
    }];
    
    [emailToName enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
        DLog(@"key: %@, value: %@", key, obj);
    }];
    
    NSMutableSet *visitedSet = [NSMutableSet set];
    NSMutableArray *ans = [NSMutableArray array];
    for(NSString *email in graph.allKeys) {
        if(![visitedSet containsObject:email]) {
            [visitedSet addObject:email];
            NSMutableArray *stack = [NSMutableArray array];
            NSMutableArray *component = [NSMutableArray array];
            [stack push:email];
            while(stack.count != 0) {
                NSString *node = (NSString *)[stack pop];
                [component addObject:node];
                for(NSString *nei in [graph objectForKey:node]) {
                    if(![visitedSet containsObject:nei]) {
                        [visitedSet addObject:nei];
                        [stack push:nei];
                    }
                }
            }
            [component sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
            [component insertObject:[emailToName objectForKey:email] atIndex:0];
            [ans addObject:component];
        }
    }
    
    return [ans copy];
}

- (void)test {
    NSArray *list1 = @[@"John", @"johnsmith@mail.com", @"john00@mail.com"];
    NSArray *list2 = @[@"John", @"johnnybravo@mail.com"];
    NSArray *list3 = @[@"John", @"johnsmith@mail.com", @"john_newyork@mail.com"];
    NSArray *list4 = @[@"Mary", @"mary@mail.com"];
    NSArray *accounts = @[list1, list2, list3, list4];
    [self accountsMerge:accounts];
    
}
@end

File: LCIntelliJ/src/LC721.java
import java.util.*;

public class LC721 {

    public static List<List<String>> accountsMerge_2(List<List<String>> accounts) {
        Map<String, String> emailToName = new HashMap<>(); // to construct answer
        Map<String, List<String>> graph = new HashMap<>();
        for (List<String> account : accounts) {
            String name = "";

            for (int i = 0; i < account.size(); i++) {
                String email = account.get(i);
                if (i == 0) {
                    name = email;
                    continue;
                }
                if (i == 1) {
                    graph.computeIfAbsent(email, x -> new ArrayList<>());
                    emailToName.put(email, name);
                    continue;
                }
                graph.computeIfAbsent(email, x -> new ArrayList<>()).add(account.get(1));
                graph.get(account.get(1)).add(email);
                emailToName.put(email, name);
            }
        }

        PrintUtils.printMapKeyStringValueList(graph);
        PrintUtils.printLine();
        PrintUtils.printMapKeyStringValueString(emailToName);

        Set<String> visited = new HashSet<>();
        List<List<String>> res = new ArrayList<>();
        for (String email : graph.keySet()) {
            if (!visited.contains(email)) {
                visited.add(email);
                Stack<String> stack = new Stack<>();
                stack.push(email);
                List<String> component = new ArrayList<>();
                while (!stack.isEmpty()) {
                    String cur = stack.pop();
                    component.add(cur);
                    for (String nei : graph.get(cur)) {

                        if (!visited.contains(nei)) {
                            stack.push(nei);
                            visited.add(nei);
                        }
                    }
                }

                Collections.sort(component);
                component.add(0, emailToName.get(email));
                res.add(component);
            }
        }
        PrintUtils.printListOfListString(res);
        PrintUtils.printLine();
        return res;
    }

    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        Map<String, String> emailToName = new HashMap<>();
        Map<String, List<String>> graph = new HashMap<>();

        for(List<String> account: accounts) {
            String name = "";
            for(int i = 0; i < account.size(); i++) {
                String email = account.get(i);
                if(i == 0) {
                    name = email;
                    continue;
                }
                if(i == 1) {
                    if(graph.containsKey(email) == false) {
                        graph.put(email, new ArrayList<String>());
                    }
                    emailToName.put(email, name);
                    continue;
                }

                if(graph.containsKey(email) == false) {
                    graph.put(email, new ArrayList<String>());
                }
                graph.get(email).add(account.get(1));
                graph.get(account.get(1)).add(email);
                emailToName.put(email, name);
            }
        }
        PrintUtils.printMapKeyStringValueList(graph);
        PrintUtils.printLine();
        PrintUtils.printMapKeyStringValueString(emailToName);

        List<List<String>> res = new ArrayList<List<String>>();
        Set<String> visited = new HashSet<>();
        for(String email : graph.keySet()) {
            if(!visited.contains(email)) {
                visited.add(email);
                Stack<String> stack = new Stack<>();
                stack.push(email);
                List<String> component = new ArrayList<>();
                while (!stack.isEmpty()) {
                    String cur = stack.pop();
                    component.add(cur);
                    for(String nei : graph.get(cur)) {
                        if(!visited.contains(nei)) {
                            stack.push(nei);
                            visited.add(nei);
                        }
                    }
                }
                Collections.sort(component);
                component.add(0, emailToName.get(email));
                res.add(component);
            }
        }
        PrintUtils.printListOfListString(res);
        PrintUtils.printLine();
        return res;
    }

    public static void main(String[] args) {
        String[] array1 = new String[]{"John", "johnsmith@mail.com", "john00@mail.com"};
        List<String> list1 = new ArrayList<>(Arrays.asList(array1));

        String[] array2 = new String[]{"John", "johnnybravo@mail.com"};
        List<String> list2 = new ArrayList<>(Arrays.asList(array2));

        String[] array3 = new String[]{"John", "johnsmith@mail.com", "john_newyork@mail.com"};
        List<String> list3 = new ArrayList<>(Arrays.asList(array3));

        String[] array4 = new String[]{"Mary", "mary@mail.com"};
        List<String> list4 = new ArrayList<>(Arrays.asList(array4));

        List<List<String>> accounts = new ArrayList<List<String>>();
        accounts.add(list1);
        accounts.add(list2);
        accounts.add(list3);
        accounts.add(list4);

        LC721 obj = new LC721();
        obj.accountsMerge(accounts);
        PrintUtils.printLine();
        obj.accountsMerge_2(accounts);
    }
}
